sidecar/proxy/envoy/manager.go

```diff
@@ -485,6 +484,12 @@ func sanitizeRules(ruleList []api.Rule) {
 					}
 				}
 			}
+
+			sum = 0
+			for j := range rule.Route.Backends {
+				sum += rule.Route.Backends[j].Weight
+				rule.Route.Backends[j].Weight = sum
+			}
```

Comment:
You can edit the filesystem with weighted_clusters as well.
Secondly, in Approach B (probability based cluster selection), it is incorrect to use the term weights. Look at the example in Envoy. To split traffic into 33/33/33, you need to set weight1 to 33, weight2 to 50 (not 66), and nothing for last case.

It just grows infinitely more complicated..

With weighted clusters, its super simple. Each cluster has the same percentage weight as specified by user. And you can adjust the weights by using the runtime_key_prefix.. E.g., 
to adjust weight of cluster1, you tweak the value of runtime_key_prefix+cluster1.